editoast/core_task/src/envs/simulation/request.rs

```diff
@@ -0,0 +1,426 @@
+use std::hash::DefaultHasher;
+use std::hash::Hash;
+use std::sync::Arc;
+
+use core_client::AsCoreRequest as _;
+use core_client::CoreClient;
+
+use crate::CoreEnv;
+use crate::Task;
+use crate::envs::pathfinding;
+
+use super::SimulationKey;
+use super::SimulationWaypointSchedule;
+
+pub(super) fn build_request(
+    core_env: &CoreEnv,
+    electrical_profiles_set_id: Option<u64>,
+    SimulationKey(sim_consist, params, pathfinding::Input(pf_consist, constraints)): &SimulationKey,
+    core_client::pathfinding::PathfindingResultSuccess {
+        path,
+        path_item_positions,
+        ..
+    }: core_client::pathfinding::PathfindingResultSuccess,
+) -> Result<core_client::simulation::Request, ()> {
+    if constraints.path_items.len() != path_item_positions.len() {
+        tracing::error!(
+            ?sim_consist,
+            ?pf_consist,
+            sim_params = ?params,
+            pf_constraints = ?constraints,
+            waypoint_count = constraints.path_items.len(),
+            computed_path_item_count = path_item_positions.len(),
+            "Core path does not respect pathfinding constraints. Please open a bug report."
+        );
+        return Err(());
+    }
+    // Thanks to SimulationTrain, all indexing from now on should be safe.
+    // If any indexing below goes wrong, it's a bug.
+
+    let mut power_restrictions = Vec::new();
+    for (range, restriction) in params.power_restrictions.iter() {
+        let from = path_item_positions[range.start];
+        let to = path_item_positions[range.end];
+        power_restrictions.push(core_client::simulation::SimulationPowerRestrictionItem {
+            from,
+            to,
+            value: restriction.clone(),
+        });
+    }
+
+    let mut margin_boundaries = Vec::<u64>::new();
+    let mut margin_values = Vec::new();
+    for (range, margin) in params.margins.iter() {
+        let from = path_item_positions[range.start];
+        let to = path_item_positions[range.end];
+        if let Some(prev) = margin_boundaries.last() {
+            debug_assert_eq!(*prev, from);
+        } else {
+            margin_boundaries.push(from);
+        }
+        margin_boundaries.push(to);
+        margin_values.push(*margin);
+    }
```

Comment:
We were 3 reviewing this piece of code, and it took us some time to figure out what this `if` was doing. I’m trying to think of a more straightforward to write this (straightforward to understand, I mean). If I’m not wrong, the following is equivalent and show the intent of pushing the first from only once, and checking for the order of values.


```suggestion
    for (range, margin) in params.margins.iter() {
        let from = path_item_positions[range.start];
        let to = path_item_positions[range.end];
        if margin_boundaries.is_empty() {
            margin_boundaries.push(from);
        }
        margin_boundaries.push(to);
        margin_values.push(*margin);
    }
    debug_assert_eq!(margin_boundaries.is_sorted(), "margin boundaries should only distributed along the path, therefore in a sorted order");
```

What do you think?